SimFastTiming/FastTimingCommon/src/BTLElectronicsSim.cc

```diff
@@ -16,30 +34,95 @@ BTLElectronicsSim::BTLElectronicsSim(const edm::ParameterSet& pset) :
 
 
 void BTLElectronicsSim::run(const mtd::MTDSimHitDataAccumulator& input,
-			    BTLDigiCollection& output) const {
-  
-  MTDSimHitData chargeColl,toa;
+			    BTLDigiCollection& output,
+			    CLHEP::HepRandomEngine *hre) const {
   
+  MTDSimHitData chargeColl, toa1, toa2;
+
   for(MTDSimHitDataAccumulator::const_iterator it=input.begin();
       it!=input.end();
       it++) {
     
-    chargeColl.fill(0.f); 
-    toa.fill(0.f);
+    chargeColl.fill(0.f);
+    toa1.fill(0.f);
+    toa2.fill(0.f);
     for(size_t i=0; i<it->second.hit_info[0].size(); i++) {
-      //time of arrival
-      float finalToA = (it->second).hit_info[1][i];
-      while(finalToA < 0.f)  finalToA+=25.f;
-      while(finalToA > 25.f) finalToA-=25.f;
-      toa[i]=finalToA;
+
+      // --- Fluctuate the total number of photo-electrons
+      float Npe = CLHEP::RandPoissonQ::shoot(hre, (it->second).hit_info[0][i]);
+      if ( Npe < EnergyThreshold_ ) continue;
+
+
+      // --- Get the time of arrival and add a channel time offset
+      float finalToA1 = (it->second).hit_info[1][i] + ChannelTimeOffset_;
+      float finalToA2 = (it->second).hit_info[1][i] + ChannelTimeOffset_;
+      
+      if (  smearChannelTimeOffset_ > 0. ){
+	float timeSmearing = CLHEP::RandGaussQ::shoot(hre, 0., smearChannelTimeOffset_);
+	finalToA1 += timeSmearing;
+	finalToA2 += timeSmearing;
+      }
+
+
+      // --- Calculate and add the time walk
+      std::array<float, 3> times = btlPulseShape_.timeAtThr(Npe/ReferencePulseNpe_,
+							    TimeThreshold1_*Npe_to_V_, 
+							    TimeThreshold2_*Npe_to_V_);
+
+
+      // --- If the pulse amplitude is smaller than TimeThreshold2, the trigger does not fire
+      if (times[1] == 0.) continue;
+
+      finalToA1 += times[0];
+      finalToA2 += times[1];
+
+
+      // --- Uncertainty due to the fluctuations of the n-th photon arrival time:
+      float sigma2_tot_thr1 = ScintillatorDecayTime_*ScintillatorDecayTime_*sigma2_pe(TimeThreshold1_,Npe);
```

Comment:
Many quantities here could be precalculated (even in the constructor) to avoid repeating calculations over and over again:
* `ScintillatorDecayTime_*ScintillatorDecayTime_`
* `SinglePhotonTimeResolution_*SinglePhotonTimeResolution_`
* `ScintillatorRiseTime_*DarkCountRate_`
* `SigmaElectronicNoise_*SigmaElectronicNoise_`
* `SigmaClock_*SigmaClock_`

then in this function, precalculate `slew*slew` and several of these quantities added in quadrature together: `(ScintillatorRiseTime_*DarkCountRate_ + SigmaElectronicNoise_*SigmaElectronicNoise_)*slew*slew + SigmaClock_*SigmaClock_` (replacing w/ precalculated quantities)

Reply:
Done